Add more stubs for cryptography

jlaine opened the pull request.

jlaine commented:
The x509 stub still needs some love, I haven't had a chance to properly represent the "Name" class.

JelleZijlstra commented on third_party/2and3/cryptography/hazmat/backends/interfaces.pyi:
```diff
@@ -0,0 +1,13 @@
+class CMACBackend: ...
```
Presumably these things have some methods? If you don't have time to stub them all out, we should add a `def __getattr__(self, name: str) -> Any` to avoid false positives for people using these classes.

jlaine replied:
The backends themselves are private API.. so it's debatable whether they should be declared at all.

srittau replied:
They are [documented](https://cryptography.io/en/latest/hazmat/backends/interfaces/), though? Also, they should all have a `metaclass=ABCMeta` base. Same is true for several of the interfaces in other files.

jlaine replied:
Would you have an example of this?

jlaine commented:
One thing I noticed is that the signature for load_pem_private_key looks wrong, I believe the password argument is bytes, not str.

jlaine commented:
CI seems to be failing because Python 2 lacks `ipaddress`, see #3309

JelleZijlstra commented:
I think the EOL of Python 2 in typeshed is probably going to be later than December 31.

srittau commented on third_party/2and3/cryptography/exceptions.pyi:
```diff
@@ -0,0 +1,2 @@
+class InvalidSignature(Exception): ...
```
Please add `__getattr__(name: str) -> Any: ...  # incomplete` to the top of this and other files that have more definitions than the stubs.

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,21 @@
+class DSAPublicKey:
+    def public_bytes(self, encoding: Encoding, format: PublicFormat) -> bytes: ...
+    def sign(self, data: bytes, padding: AsymmetricPadding, algorithm: HashAlgorithm) -> bytes: ...
+    def verify(self, signature: bytes, data: bytes, padding: AsymmetricPadding, algorithm: HashAlgorithm) -> None: ...
```
These methods should also be decorated using `@abstractmethod`.

jlaine commented:
@srittau I have added a lot more stubs, feel free to review

srittau commented on third_party/2and3/cryptography/hazmat/bindings/openssl/binding.pyi:
```diff
@@ -0,0 +1,6 @@
+from typing import Any
+
+class Binding:
```
```suggestion
class Binding(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/bindings/openssl/binding.pyi:
```diff
@@ -0,0 +1,6 @@
+
+class Binding:
+    ffi: Any
+    lib: Any
```
You could use `Optional[Any]` here, since `None` is definitely one of the possible values. (Import required.)

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+
+DHParametersWithSerialization = DHParameters
+
+class DHParameterNumbers:
```
```suggestion
class DHParameterNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+    @abstractmethod
+    def private_numbers(self) -> DHPrivateNumbers: ...
+
+class DHPrivateNumbers:
```
```suggestion
class DHPrivateNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+class DHParameterNumbers:
+    p: int
+    g: int
+    q: int
```
These three are using `cryptography.utils.read_only_property`, which is only a fancy way of using `@property`, so we could just do that here:

```python
@property
def p(self) -> int: ...
@property
def g(self) -> int: ...
@property
def q(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+
+class DHPrivateNumbers:
+    public_numbers: DHPublicNumbers
+    x: int
```
Also properties.

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+    def private_key(self, backend: DHBackend) -> DHPrivateKey: ...
+
+class DHPublicKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+
+DHPublicKeyWithSerialization = DHPublicKey
+
+class DHPublicNumbers:
```
```suggestion
class DHPublicNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dh.pyi:
```diff
@@ -0,0 +1,68 @@
+
+class DHPublicNumbers:
+    parameter_numbers: DHParameterNumbers
+    y: int
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+    @abstractmethod
+    def parameter_numbers(self) -> DSAParameterNumbers: ...
+
+class DSAParameterNumbers:
```
```suggestion
class DSAParameterNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+class DSAParameterNumbers:
+    p: int
+    q: int
+    g: int
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+    def parameters(self, backend: DSABackend) -> DSAParameters: ...
+
+class DSAPrivateKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+    @abstractmethod
+    def private_numbers(self) -> DSAPrivateNumbers: ...
+
+class DSAPrivateNumbers:
```
```suggestion
class DSAPrivateNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+
+class DSAPrivateNumbers:
+    x: int
+    public_numbers: DSAPublicNumbers
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+    def __init__(self, x: int, public_numbers: DSAPublicNumbers) -> None: ...
+
+class DSAPublicKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+
+DSAPublicKeyWithSerialization = DSAPublicKey
+
+class DSAPublicNumbers:
```
```suggestion
class DSAPublicNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/dsa.pyi:
```diff
@@ -1,4 +1,64 @@
+
+class DSAPublicNumbers:
+    y: int
+    parameter_numbers: DSAParameterNumbers
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+
+class EllipticCurve(metaclass=ABCMeta):
+    key_size: int
+    name: str
```
Abstract properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+class SECT571K1(EllipticCurve): ...
+class SECT571R1(EllipticCurve): ...
+
+class EllipticCurveOID:
```
```suggestion
class EllipticCurveOID(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+    SECT409K1: ObjectIdentifier
+    SECT409R1: ObjectIdentifier
+    SECT571K1: ObjectIdentifier
+    SECT571R1: ObjectIdentifier
```
```python
    FOO: ClassVar[ObjectIdentifier]
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+    SECT571R1: ObjectIdentifier
+
+class EllipticCurvePrivateKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+    @abstractmethod
+    def exchange(self, algorithm: ECDH, peer_public_key: EllipticCurvePublicKey) -> bytes: ...
+    @abstractmethod
+    def public_key(self) -> EllipticCurvePublicKey: ...
```
This also has a `curve` abstract property.

jlaine replied:
If it does, it doesn't seem to be documented: https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ec/#cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePrivateKey

jlaine replied:
I'll file an issue against `cryptography`, it's missing from the docs

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+    @abstractmethod
+    def private_numbers(self) -> EllipticCurvePrivateNumbers: ...
+
+class EllipticCurvePrivateNumbers:
```
```suggestion
class EllipticCurvePrivateNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+
+class EllipticCurvePrivateNumbers:
+    private_value: int
+    public_numbers: EllipticCurvePublicNumbers
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+
+class EllipticCurvePublicKey(metaclass=ABCMeta):
+    curve: EllipticCurve
+    key_size: int
```
Abstract properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+class EllipticCurvePublicNumbers:
+    curve: EllipticCurve
+    x: int
+    y: int
```
Properties

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+
+EllipticCurvePublicKeyWithSerialization = EllipticCurvePublicKey
+
+class EllipticCurvePublicNumbers:
```
```suggestion
class EllipticCurvePublicNumbers(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+class EllipticCurveSignatureAlgorithm(metaclass=ABCMeta):
+    algorithm: Union[HashAlgorithm, Prehashed]
+
+class ECDH: ...
```
```suggestion
class ECDH(object): ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/ec.pyi:
```diff
@@ -0,0 +1,112 @@
+    def public_key(self, backend: EllipticCurveBackend) -> EllipticCurvePublicKey: ...
+
+class EllipticCurveSignatureAlgorithm(metaclass=ABCMeta):
+    algorithm: Union[HashAlgorithm, Prehashed]
```
```suggestion
    @property
    @abstractmethod
    def algorithm(self) -> Union[HashAlgorithm, Prehashed]: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+
+from cryptography.hazmat.primitives.hashes import HashAlgorithm
+
+class AsymmetricPadding:
```
```suggestion
class AsymmetricPadding(metaclass=ABCMeta):
```

jlaine replied:
Is this correct, since there are no abstract methods? (there is only the "abstract" name property, but I believe abstract properties aren't supported)

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+from cryptography.hazmat.primitives.hashes import HashAlgorithm
+
+class AsymmetricPadding:
+    name: str
```
```suggestion
    @property
    @abstractmethod
    def name(self) -> str: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+class AsymmetricPadding:
+    name: str
+
+class MGF1:
```
```suggestion
class MGF1(object):
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+    name: str
+
+class MGF1:
+    def __init__(self, algorithm: HashAlgorithm): ...
```
```suggestion
    def __init__(self, algorithm: HashAlgorithm) -> None: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+    def __init__(self, algorithm: HashAlgorithm): ...
+
+class OAEP(AsymmetricPadding):
+    def __init__(self, mgf: MGF1, algorithm: HashAlgorithm, label: Optional[bytes]): ...
```
```suggestion
    def __init__(self, mgf: MGF1, algorithm: HashAlgorithm, label: Optional[bytes]) -> None: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+class PKCS1v15(AsymmetricPadding): ...
+
+class PSS(AsymmetricPadding):
+    MAX_LENGTH: int
```
```suggestion
    MAX_LENGTH: ClassVar[object]
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/padding.pyi:
```diff
@@ -0,0 +1,18 @@
+
+class PSS(AsymmetricPadding):
+    MAX_LENGTH: int
+    def __init__(self, mgf: MGF1, salt_length: int): ...
```
```suggestion
    def __init__(self, mgf: MGF1, salt_length: int) -> None: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/rsa.pyi:
```diff
@@ -1,31 +1,45 @@
-    def signer(self, padding, algorithm): ...
-    def decrypt(self, ciphertext: bytes, padding) -> bytes: ...
+class RSAPrivateKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/rsa.pyi:
```diff
@@ -1,31 +1,45 @@
-    def key_size(self) -> int: ...
-    def public_numbers(self) -> RSAPublicNumbers: ...
+class RSAPublicKey(metaclass=ABCMeta):
+    key_size: int
```
```suggestion
    @property
    @abstractmethod
    def key_size(self) -> int: ...
```

srittau commented on third_party/2and3/cryptography/hazmat/primitives/asymmetric/utils.pyi:
```diff
@@ -0,0 +1,6 @@
+def decode_dss_signature(signature: bytes) -> Tuple[int, int]: ...
+def encode_dss_signature(r: int, s: int) -> bytes: ...
+
+class Prehashed: ...
```
```suggestion
class Prehashed(object): ...
```

jlaine commented on third_party/2and3/cryptography/exceptions.pyi:
```diff
@@ -0,0 +1,6 @@
+class InvalidSignature(Exception): ...
+class InvalidTag(Exception): ...
+class NotYetFinalized(Exception): ...
+class UnsupportedAlgorithm(Exception): ...
```
Definitely, I had missed that one.

jlaine commented on third_party/2and3/cryptography/fernet.pyi:
```diff
@@ -0,0 +1,15 @@
+from typing import List, Optional
+
+class Fernet:
```
Ah, good old Python 2. There are going to be a lot of these :(

jlaine commented:
Thanks so much for taking the time for this first review, I've addressed all the issues you pointed out. The changes related to abstract properties seem to have broken CI though, it looks as though I need to go over all the subclasses to add the corresponding concrete properties.

jlaine commented on third_party/2and3/cryptography/hazmat/backends/interfaces.pyi:
```diff
@@ -0,0 +1,196 @@
+)
+
+class CipherBackend(metaclass=ABCMeta):
+    @abstractmethod
```
That was my opinion too, but as @srittau pointed out in https://github.com/python/typeshed/pull/3307#discussion_r331999899 these methods are part of the `cryptography` documentation.. so public API?

alex commented:
We're unlikely to be convinced to bring these in-tree for cryptography until we're python3 only and can use the native syntax.

srittau commented:
@alex I'd just like to point out that the syntax of stubs is independent of the supported Python version. Even when checking Python 2.7 code, type checkers will support stub files using newer features.

srittau commented on third_party/2and3/cryptography/hazmat/primitives/ciphers/modes.pyi:
```diff
@@ -0,0 +1,52 @@
+    nonce: bytes
+
+class ModeWithTweak(metaclass=ABCMeta):
+    tweak: bytes
```
This field and all fields above are abstract properties. (Probably affects fields of abstract classes below, too, but I didn't check.)

alex commented:
Yes, I understand the stubs syntax works in both Python2/3. We don't plan to adopt type hints until they can be done directly inline in the source code.

srittau commented:
Seems sensible. We are happy to have them in typeshed as long as required.

srittau commented:
Looks good, I will merge when the build failure is fixed.

jlaine commented:
Let's see if CI completes this time. I caught a couple of other places in `hazmat.ciphers.algorithms` where abstract properties were needed instead of members.

jlaine commented:
Thanks for all your guidance @srittau, this is one friendly project!
